File: CivitasJava/civitas/CivitasJuego.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;
import java.util.logging.Level;
import java.util.logging.Logger;

public class CivitasJuego {
    
    // Atributos
    private int indiceJugadorActual;
    MazoSorpresas mazo;
    Tablero tablero;
    ArrayList<Jugador> jugadores;
    EstadosJuego estado;
    GestorEstados gestorEstados;
    
    // Métodos
   
    private void avanzaJugador(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
	int posicionActual = jugadorActual.getNumCasillaActual();
	int tirada = Dado.getInstance().tirar();
	int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
	Casilla casilla = tablero.getCasilla(posicionNueva);
	this.contabilizarPasosPorSalida(jugadorActual);
	jugadorActual.moverACasilla(posicionNueva);
        try {
            casilla.recibeJugador(indiceJugadorActual, jugadores);
        } catch (Exception ex){}
	this.contabilizarPasosPorSalida(jugadorActual);
    }
    
    public boolean cancelarHipoteca(int ip){
        return getJugadorActual().cancelarHipoteca(ip);
    }
    
    public CivitasJuego(ArrayList<String> nombres){
        int numJugadores = nombres.size();
        
        // Inicializa los jugadores
        jugadores = new ArrayList<>(numJugadores);
        for (int i = 0; i < numJugadores; i++)
            jugadores.add(new Jugador(nombres.get(i)) );
        
        // Inicializa el gestor de estados y el estado inicial
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        
        // Decide qué jugador empieza
        indiceJugadorActual = Dado.getInstance().quienEmpieza(numJugadores);
        
        // Inicializa el tablero y el mazo
        mazo = new MazoSorpresas();
        inicializarTablero(mazo);
        inicializarMazoSorpresas(tablero);    
    }
    
    public boolean comprar(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        Casilla casilla = tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo = ((CasillaCalle) casilla).getTituloPropiedad();
        return jugadorActual.comprar(titulo);
    }
    
    /**
     * Método para construir una casa en una propiedad del jugador
     * @param ip índice de la propiedad en la lista de propiedades del jugador
     * @return true si la operación tiene éxito
     */
    public boolean construirCasa(int ip){
        return getJugadorActual().construirCasa(ip);
    }
    
    /**
     * Método para construir un hotel en una propiedad del jugador
     * @param ip índice de la propiedad en la lista de propiedades del jugador
     * @return true si la operación tiene éxito
     */
    public boolean construirHotel(int ip){
        return getJugadorActual().construirHotel(ip);
    }
    
    /**
     * Método para tramitar las recompensas de pasar por salida de un jugador
     * @param jugadorActual
     */
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        while(tablero.getPorSalida() > 0){
            jugadorActual.pasaPorSalida();
        };
    }
    
    public boolean finalDelJuego(){        
        for (int i = 0; i < jugadores.size(); i++)
            if (jugadores.get(i).enBancarrota())
                return true;
        
        return false;
    }
    
    public Casilla getCasillaActual(){
        int numCasilla = getJugadorActual().getNumCasillaActual();
        return tablero.getCasilla(numCasilla);
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    public boolean hipotecar(int ip){
        return getJugadorActual().hipotecar(ip);
    }
    
    public String infoJugadorTexto(){
        String str = "Jugador actual: " + getJugadorActual().getNombre() 
                   + " en la casilla: " + getJugadorActual().getNumCasillaActual();
                
        return str;
    }
    
    /**
     * Método para inicializar el mazo de sorpresas del juego
     * @param tablero de juego
     */
    private void inicializarMazoSorpresas(Tablero tablero){
        mazo.alMazo(new SorpresaSalirCarcel(mazo, "Salvoconducto"));
        mazo.alMazo(new SorpresaPorJugador(-300, "PAGAR_PORJUGADOR_1"));
        mazo.alMazo(new SorpresaPorJugador(300, "COBRAR_PORJUGADOR_1"));
        mazo.alMazo(new SorpresaIrCasilla(tablero, 9, "Ir a casilla 9"));
        mazo.alMazo(new SorpresaIrCasilla(tablero, 19, "Ir a casilla 19"));
        mazo.alMazo(new SorpresaIrCarcel(tablero, tablero.getCarcel(), "Ir a la cárcel"));
        mazo.alMazo(new SorpresaPorCasaHotel(-100, "Pagar 100 por cada casa/hotel"));
        mazo.alMazo(new SorpresaPorCasaHotel(100, "Cobrar 100 por cada casa/hotel"));
        mazo.alMazo(new SorpresaPagarCobrar(500, "Cobrar 500"));
        mazo.alMazo(new SorpresaPagarCobrar(-500, "Pagar 500"));
    }
    
    /**
     * Méteodo para inicializar el tablero de juego
     * @param mazo mazo de sorpresas del juego 
     */
    private void inicializarTablero(MazoSorpresas mazo){
        int numCasillaCarcel = 5;
        tablero = new Tablero(numCasillaCarcel);
        
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle1", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle2", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa1"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle3", 10, 1.1f, 500, 600, 250)));
        // Cárcel
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle4", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle5", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa2"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle6", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new Casilla("PARKING"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle7", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle8", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa3"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle9", 10, 1.1f, 500, 600, 250)));
        tablero.añadeJuez();
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle10", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle11", 10, 1.1f, 500, 600, 250)));
        tablero.añadeCasilla(new CasillaImpuesto(10, "IMPUESTO"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle12", 10, 1.1f, 500, 600, 250)));
    }
    
    /**
     * Método para pasar el turno de un jugador a otro
     */
    private void pasarTurno(){
        if (indiceJugadorActual < (jugadores.size() -1) )
            indiceJugadorActual++;
        else
            indiceJugadorActual = 0;
    }
    
    public ArrayList<Jugador> ranking(){
        ArrayList<Jugador> ranking = new ArrayList<>(jugadores);
        Collections.sort(ranking, Jugador::compareTo);
        return ranking;
    }
    
    public boolean salirCarcelPagando(){
        return getJugadorActual().salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return getJugadorActual().salirCarcelTirando();
    }
    
	/**
	 * @return OperacionesJuego siguienteOperacion
	 */
    public OperacionesJuego siguientePaso(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
		OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
		
		if (operacion == OperacionesJuego.PASAR_TURNO){
			pasarTurno();
			siguientePasoCompletado(operacion);
		} else if (operacion == OperacionesJuego.AVANZAR){
			avanzaJugador();
			siguientePasoCompletado(operacion);
		}
		
		return operacion;
    }

    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorEstados.siguienteEstado(getJugadorActual(), estado, operacion);
    }
    
    public boolean vender(int ip){
        return getJugadorActual().vender(ip);
    }
}

File: CivitasRuby/civitas_juego.rb
# encoding:utf-8
module Civitas
  
require_relative 'gestor_estados'  
require_relative 'jugador'
require_relative 'mazo_sorpresas'
require_relative 'sorpresa'
require_relative 'dado'
require_relative 'tablero'
require_relative 'titulo_propiedad'
require_relative 'operaciones_juego'
require_relative 'casilla_calle'
require_relative 'casilla_impuesto'
require_relative 'casilla_juez'
require_relative 'casilla_sorpresa'
require_relative 'sorpresa_ir_carcel'
require_relative 'sorpresa_ir_casilla'
require_relative 'sorpresa_pagar_cobrar'
require_relative 'sorpresa_por_casa_hotel'
require_relative 'sorpresa_por_jugador'
require_relative 'sorpresa_salir_carcel'
  
class CivitasJuego
  
  def initialize(nombres)
    @jugadores = []
    num_jugadores = nombres.size
    for i in 0..num_jugadores-1
      @jugadores.push(Jugador.new(nombres[i]))
    end
      
    @gestorEstados = Gestor_estados.new
    @estado = @gestorEstados.estado_inicial
    @indiceJugadorActual = Dado.instance.quien_empieza(num_jugadores)
    
    @mazo = MazoSorpresas.new
    inicializar_tablero(@mazo)
    inicializar_mazo_sorpresas(@tablero)
  end
  
  def avanza_jugador
    jugador_actual = @jugadores[@indiceJugadorActual]
		posicion_actual = jugador_actual.numCasillaActual
		tirada = Dado.instance.tirar()
		posicion_nueva = @tablero.nueva_posicion(posicion_actual, tirada)
		casilla = @tablero.get_casilla(posicion_nueva)
		contabilizar_pasos_por_salida(jugador_actual)
		jugador_actual.mover_a_casilla(posicion_nueva)
		casilla.recibe_jugador(@indiceJugadorActual, @jugadores)
		contabilizar_pasos_por_salida(jugador_actual)
  end
  
  def cancelar_hipoteca(ip)
    get_jugador_actual.cancelar_hipoteca(ip)
  end
  
  def comprar
    jugador_actual = @jugadores[@indiceJugadorActual]
    casilla = @tablero.get_casilla(jugador_actual.numCasillaActual)
    titulo = casilla.tituloPropiedad
    jugador_actual.comprar(titulo)
  end
  
  def construir_casa(ip)
    get_jugador_actual.construir_casa(ip)
  end
  
  def construir_hotel(ip)
    get_jugador_actual.construir_hotel(ip)
  end
  
  def contabilizar_pasos_por_salida(jugador)
    while @tablero.get_por_salida > 0 do
      jugador.pasa_por_salida
    end
  end
  
  def final_del_juego 
    for i in 0..@jugadores.size-1
      if @jugadores[i].en_bancarrota
        return true
      end
    end
    
    return false
  end
  
  def get_casilla_actual
    num_casilla = get_jugador_actual.numCasillaActual
    @tablero.get_casilla(num_casilla)
  end
  
  def get_jugador_actual
    return @jugadores[@indiceJugadorActual]
  end
  
  def hipotecar(ip)
    get_jugador_actual.hipotecar(ip)
  end
  
  def info_jugador_texto
    str = "Jugador actual: " + get_jugador_actual.nombre + " en la casilla: " + get_jugador_actual.numCasillaActual.to_s
    return str
  end
  
  def inicializar_mazo_sorpresas(tablero)
    @mazo.al_mazo( SorpresaIrCarcel.new(tablero, tablero.numCasillaCarcel, "Ir a la carcel") )
    @mazo.al_mazo( SorpresaSalirCarcel.new(@mazo, "Salvoconducto") )
    @mazo.al_mazo( SorpresaPorJugador.new( -300, "Pagas 300 a cada jugador") )
    @mazo.al_mazo( SorpresaPorJugador.new( 300, "Cobras 300 por cada jugador" ) )
    @mazo.al_mazo( SorpresaIrCasilla.new(tablero, 9, "Ir a la casilla 9"))
    @mazo.al_mazo( SorpresaIrCasilla.new(tablero, 19, "Ir a la casilla 19"))
    @mazo.al_mazo( SorpresaPorCasaHotel.new(-100, "Pagas 100 por cada casas u hotel de tu propiedad") )
    @mazo.al_mazo( SorpresaPorCasaHotel.new(100, "Cobras 100 por cada casa u hotel de tu propiedad") )
    @mazo.al_mazo( SorpresaPagarCobrar.new(500, "Cobras 500") )
    @mazo.al_mazo( SorpresaPagarCobrar.new(-500, "Pagas 500") )
  end
  
  def inicializar_tablero(mazo)
    num_casilla_carcel = 5
    @tablero = Tablero.new(num_casilla_carcel)
    
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle1", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle2", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaSorpresa.new(mazo, "Sorpresa1"))
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle3", 10, 1.1, 500, 600, 250)) )
    #Cárcel
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle4", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle5", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaSorpresa.new(mazo, "Sorpresa2"))
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle6", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( Casilla.new("PARKING") )
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle7", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle8", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaSorpresa.new(mazo, "Sorpresa3"))
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle9", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_juez
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle10", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle11", 10, 1.1, 500, 600, 250)) )
    @tablero.añade_casilla( CasillaImpuesto.new(10, "IMPUESTO") )
    @tablero.añade_casilla( CasillaCalle.new(TituloPropiedad.new("Calle12", 10, 1.1, 500, 600, 250)) )
  end
  
  def pasar_turno
    if @indiceJugadorActual < @jugadores.length-1
      @indiceJugadorActual += 1
    else
      @indiceJugadorActual = 0
    end
  end
  
  def ranking
    ranking = @jugadores[0 .. @jugadores.length-1]
    ranking.sort.reverse
    return ranking
  end
  
  def salir_carcel_pagando
    get_jugador_actual.salir_carcel_pagando
  end
  
  def salir_carcel_tirando
    get_jugador_actual.salir_carcel_tirando
  end
  
  def siguiente_paso
    jugador_actual = @jugadores[@indiceJugadorActual]
    operacion = @gestorEstados.operaciones_permitidas(jugador_actual, @estado)
    if operacion == Operaciones_juego::PASAR_TURNO
      pasar_turno
      siguiente_paso_completado(operacion)
    elsif operacion == Operaciones_juego::AVANZAR
      avanza_jugador
      siguiente_paso_completado(operacion)
    end
    
    operacion
  end
  
  def siguiente_paso_completado(operacion)
    @estado = @gestorEstados.siguiente_estado(get_jugador_actual, @estado, operacion)
  end
  
  def vender(ip)
    get_jugador_actual.vender(ip)
  end

  attr_reader :numCasillaActual
end
end
